Model/Api/Tax.php

```diff
@@ -169,17 +181,58 @@ public function createInstorePickOption($totalsInformation, $currencyCode, $ship
         $storeAddress->setRegion($ship_to_store_option['address']['region']);
         $storeAddress->setPostalCode($ship_to_store_option['address']['postal_code']);
         
+        $ship_to_store_option = $this->getShippingDiscount($totalsInformation, $currencyCode, $ship_to_store_option);
+        
         $shipToStoreOption = $this->shipToStoreOptionFactory->create();
         $shipToStoreOption->setReference($ship_to_store_option['reference']);
-        $shipToStoreOption->setCost(CurrencyUtils::toMinor($totalsInformation->getShippingAmount(), $currencyCode));
+        $shipToStoreOption->setCost($ship_to_store_option['cost']);
         $shipToStoreOption->setStoreName($ship_to_store_option['store_name']);
         $shipToStoreOption->setAddress($storeAddress);
         $shipToStoreOption->setDistance($ship_to_store_option['distance']);
         $shipToStoreOption->setDistanceUnit($ship_to_store_option['distance_unit']);
-        $shipToStoreOption->setTaxAmount(CurrencyUtils::toMinor($totalsInformation->getShippingTaxAmount(), $currencyCode));
+        $shipToStoreOption->setTaxAmount($ship_to_store_option['tax_amount']);
 
         return $shipToStoreOption;
     }
+    
+    /**
+     * @param TotalsInterface $totalsInformation
+     * @param string $currencyCode
+     * @param array $shippingOption
+     * @return array
+     * @throws \Exception
+     */
+    protected function getShippingDiscount($totalsInformation, $currencyCode, $shippingOption)
+    {
+        $shippingDiscountAmount = $this->eventsForThirdPartyModules->runFilter(
+            "collectShippingDiscounts",
+            $totalsInformation->getShippingDiscountAmount(),
+            $this->quote,
+            $this->quote->getShippingAddress()
+        );
+        if ($shippingDiscountAmount >= DiscountHelper::MIN_NONZERO_VALUE) {
+            $service = $shippingOption['service'] ?? '';
+            $shippingCost = $totalsInformation->getShippingAmount() - $shippingDiscountAmount;
+            $shippingRoundedCost = CurrencyUtils::toMinor($shippingCost, $currencyCode);
+            $diff = CurrencyUtils::toMinorWithoutRounding($shippingCost, $currencyCode) - $shippingRoundedCost;
+            $taxAmount = round(CurrencyUtils::toMinorWithoutRounding($totalsInformation->getShippingTaxAmount(), $currencyCode) + $diff);
+            if ($shippingRoundedCost == 0) {
+                $service .= ' [free&nbsp;shipping&nbsp;discount]';
+            } else {
+                $discount = $this->priceHelper->currency($shippingDiscountAmount, true, false);
+                $service .= " [$discount" . "&nbsp;discount]";
+            }
+            $shippingOption['service'] = html_entity_decode($service);
```

Comment:
```suggestion
        $shipping_option['service'] = sprintf(
            '%s [%s discount]',
            html_entity_decode($service),
            $shippingRoundedCost == 0
                ? 'free shipping'
                : $this->priceHelper->currency($shippingDiscountAmount, true, false)
        );
```
I don't remember why that was added, perhaps regarding something passed in the service title, not the whitespace. In that case, the above suggestion looks better. However, as we are not sure, you better leave it as is. Or test it with some live shipping options (not free and flat), provided out of the box in Magento (Fedex, UPS, USPS, something should work for testing without keys).

Reply:
I prefer to leave it as-is, cause it works as expected so far.